### PR TITLE
Lock collateral and return debt handlers

### LINKED ISSUE #2: Implement lock ATOM message handler

### LINKED ISSUE #4: Implement return MEI message handler

### PR DESCRIPTION
**Pull request by tharamalai**

### CODE REVIEW

File: x/meicdp/client/cli/tx.go [hunk 1]
```diff
@@ -54,8 +58,8 @@ func GetCmdRequest(cdc *codec.Codec) *cobra.Command {
 		Long: strings.TrimSpace(
 			fmt.Sprintf(`Make a new request via an existing oracle script with the configuration flags.
 Example:
-$ %s tx consuming request 1 -c 1234abcdef -r 4 -v 3 -x 20 -w 50 -g 5000 --from mykey
-$ %s tx consuming request 1 --calldata 1234abcdef --requested-validator-count 4 --sufficient-validator-count 3 --expiration 20 --prepare-gas 50 --execute-gas 5000 --from mykey
+$ %s tx meichain request 1 -c 1234abcdef -r 4 -v 3 -x 20 -w 50 -g 5000 --from mykey
```
Comment (defi-pizza) [inline x/meicdp/client/cli/tx.go]: Remove!

File: x/meicdp/handler.go [hunk 2]
```diff
@@ -54,12 +54,59 @@ func NewHandler(keeper Keeper) sdk.Handler {
...
+}
+
+func handleMsgReturnDebt(ctx sdk.Context, keeper Keeper, msg types.MsgReturnDebt) (*sdk.Result, error) {
+
+	cdp := keeper.GetCDP(ctx, msg.Sender)
+
+	// Subtract debt on CDP
+	newDebt := cdp.DebtAmount.Sub(msg.Amount)
+	cdp.DebtAmount = newDebt
+
```
Comment (defi-pizza) [inline x/meicdp/handler.go:103]: Add `// TODO: Pay fee`

**Review by defi-pizza — CHANGES_REQUESTED** [review]
Let's make all these messages take `uint64` instead of `sdk.Coins` so we don't need to worry about wrong denoms.

File: x/meicdp/client/cli/tx.go (hunk 1 above)
Comment (tharamalai) [reply x/meicdp/client/cli/tx.go]: fixed

File: x/meicdp/handler.go (hunk 2 above)
Comment (tharamalai) [reply x/meicdp/handler.go:103]: fixed

File: x/meicdp/client/cli/tx.go [hunk 3]
```diff
@@ -39,23 +38,25 @@ func GetTxCmd(storeKey string, cdc *codec.Codec) *cobra.Command {
...
+		Use:   "lock [amount]",
+		Short: "Lock collateral to the CDP.",
 		Args:  cobra.ExactArgs(1),
 		Long: strings.TrimSpace(
-			fmt.Sprintf(`Make a new request via an existing oracle script with the configuration flags.
+			fmt.Sprintf(`Lock collateral to the CDP.
 Example:
-$ %s tx consuming request 1 -c 1234abcdef -r 4 -v 3 -x 20 -w 50 -g 5000 --from mykey
-$ %s tx consuming request 1 --calldata 1234abcdef --requested-validator-count 4 --sufficient-validator-count 3 --expiration 20 --prepare-gas 50 --execute-gas 5000 --from mykey
+$ %s tx meicdp lock 100000uatom
```
Comment (defi-pizza) [inline x/meicdp/client/cli/tx.go]: update example?

File: x/meicdp/handler.go [hunk 4]
```diff
@@ -17,49 +15,71 @@ func NewHandler(keeper Keeper) sdk.Handler {
...
+	return &sdk.Result{}, nil
+}
+
+func handleMsgReturnDebt(ctx sdk.Context, keeper Keeper, msg types.MsgReturnDebt) (*sdk.Result, error) {
+
+	cdp := keeper.GetCDP(ctx, msg.Sender)
+
+	// Subtract debt on CDP
+	returnAmount := sdk.NewCoin(types.AtomUnit, sdk.NewInt(int64(msg.Amount)))
```
Comment (defi-pizza) [inline x/meicdp/handler.go]: Mei?

File: x/meicdp/handler.go [hunk 5]
```diff
@@ -17,49 +15,71 @@ func NewHandler(keeper Keeper) sdk.Handler {
...
+	returnAmountCoins := sdk.NewCoins(returnAmount)
+
+	// new debt
+	cdp.DebtAmount = cdp.DebtAmount - msg.Amount
```
Comment (defi-pizza) [inline x/meicdp/handler.go]: Careful of overflow

**Review by defi-pizza — CHANGES_REQUESTED** [review]

File: x/meicdp/types/msgs.go [hunk 6]
```diff
@@ -42,7 +42,7 @@ func NewMsgRequestData(
 func (msg MsgRequestData) Route() string { return RouterKey }
 
 // Type implements the sdk.Msg interface for MsgRequestData.
-func (msg MsgRequestData) Type() string { return "consuming" }
+func (msg MsgRequestData) Type() string { return "meichain" }
```
Comment (cryptomania49) [inline x/meicdp/types/msgs.go]: It should be "request" the old code also wrong

File: x/meicdp/handler.go (hunk 4 above)
Comment (tharamalai) [reply x/meicdp/handler.go]: fixed

File: x/meicdp/alias.go [hunk 7]
```diff
@@ -18,6 +18,5 @@ var (
 )
 
 type (
-	Keeper         = keeper.Keeper
-	MsgRequestData = types.MsgRequestData
+	Keeper = keeper.Keeper
```
Comment (cryptomania49) [inline x/meicdp/alias.go:21]: Can u add msgs

File: x/meicdp/handler.go [hunk 8]
```diff
@@ -17,49 +15,71 @@ func NewHandler(keeper Keeper) sdk.Handler {
...
+	// Store CDP
+	keeper.SetCDP(ctx, cdp)
+
+	// Transfer Mei from user to CDP. Transaction fails if sender's balance is insufficient.
+	moduleAddress := types.GetMeiCDPAddress()
+	err := keeper.BankKeeper.SendCoins(ctx, msg.Sender, moduleAddress, returnAmountCoins)
+	if err != nil {
+		return nil, err
+	}
+
```
Comment (cryptomania49) [inline x/meicdp/handler.go:83]: Burn Mei?

File: x/meicdp/client/cli/tx.go (hunk 3 above)
Comment (tharamalai) [reply x/meicdp/client/cli/tx.go]: Removed request tx

File: x/meicdp/handler.go (hunk 5 above)
Comment (tharamalai) [reply x/meicdp/handler.go]: Used Int instead of uint64

File: x/meicdp/types/msgs.go (hunk 6 above)
Comment (tharamalai) [reply x/meicdp/types/msgs.go]: Removed MsgRequestData

File: x/meicdp/handler.go (hunk 8 above)
Comment (tharamalai) [reply x/meicdp/handler.go:112]: Added burn Mei

File: x/meicdp/alias.go (hunk 7 above)
Comment (tharamalai) [reply x/meicdp/alias.go]: Added messages

File: x/meicdp/types/key.go [hunk 9]
```diff
@@ -60,3 +80,8 @@ func uint64ToBytes(num uint64) []byte {
 	binary.BigEndian.PutUint64(result, num)
 	return result
 }
+
+// GetMeiCDPAddress returns Mei CDP Address
+func GetMeiCDPAddress() sdk.AccAddress {
```
[the remark this answers is not in the document]
Comment (tharamalai) [reply x/meicdp/types/key.go]: Removed

File: x/meicdp/client/cli/tx.go [hunk 10]
```diff
@@ -39,23 +38,26 @@ func GetTxCmd(storeKey string, cdc *codec.Codec) *cobra.Command {
...
```
[the remark this answers is not in the document]
Comment (tharamalai) [reply x/meicdp/client/cli/tx.go]: Updated comment

File: x/meicdp/types/msgs.go [hunk 11]
```diff
@@ -100,15 +26,15 @@ func NewMsgBorrowDebt(
 func (msg MsgBorrowDebt) Route() string { return RouterKey }
 
 // Type implements the sdk.Msg interface for MsgBorrowDebt.
-func (msg MsgBorrowDebt) Type() string { return "borrow_debt" }
+func (msg MsgBorrowDebt) Type() string { return "meichain/borrow_debt" }
```
[the remark this answers is not in the document]
Comment (tharamalai) [reply x/meicdp/types/msgs.go]: Removed

**Pull request merged**